docs/single_member_restoration.md

```diff
@@ -0,0 +1,16 @@
+## Restoration of a single member in multi-node etcd deployed by etcd-druid.
+
+**Note**:
+- For a cluster with n members, we are proposing the solution to only single member restoration within a etcd cluster not the quorum loss scenario (when majority of members within a cluster fail).
+- In this proposal we are not targetting the recovery of single member which got separated from cluster due to [network partition](https://etcd.io/docs/v3.3/op-guide/failures/#network-partition).
+
+## Motivation
+If a single etcd member within a multi-node etcd cluster goes down due to DB corruption/PVC corruption/Invalid data-dir then it needs to be brought back. Unlike in the single-node case, a minority member of a multi-node cluster can't be restored from the snapshots present in storage container as you can't restore from the old snapshots as it contains the metadata information of cluster which leads to **memberID mismatch** that prevents the new member from coming up as new member is getting it's metadata information from db which got restore from old snapshots.
+
+## Solution
+- Corresponding backup-restore sidecar if detects that its corresponding etcd is down due to [data-dir corruption](https://github.com/gardener/etcd-backup-restore/blob/7d27a47f5793b0949492d225ada5fd8344b6b6a2/pkg/initializer/validator/datavalidator.go#L177) or [Invalid data-dir](https://github.com/gardener/etcd-backup-restore/blob/7d27a47f5793b0949492d225ada5fd8344b6b6a2/pkg/initializer/validator/datavalidator.go#L204)
+- Then, if it's a cluster with cluster size `>1` then backup-restore will first remove the etcd failed member using the [MemberRemove API](https://github.com/etcd-io/etcd/blob/ae9734ed278b7a1a7dfc82e800471ebbf9fce56f/clientv3/cluster.go#L45-L46) call and clean the data-dir of failed etcd member.
+- It won't affect the etcd cluster as quorum is still maintained.
+- After successfully removing failed member from the cluster, backup-restore sidecar will try to add a new member in a cluster to get the same cluster size as before.
+- Backup-restore firstly adds new member as a [Learner](https://etcd.io/docs/v3.3/learning/learner/) using the [MemberAddAsLearner API](https://github.com/etcd-io/etcd/blob/ae9734ed278b7a1a7dfc82e800471ebbf9fce56f/clientv3/cluster.go#L42-L43) call, once learner was added to the cluster and it's get in sync with leader and becomes up-to-date then promote the learner to a voting member using [MemberPromote API](https://github.com/etcd-io/etcd/blob/ae9734ed278b7a1a7dfc82e800471ebbf9fce56f/clientv3/cluster.go#L51-L52) call.
```

Comment:
Is it possible that backup-restore container fails to promote a member even after several retries? At what point a corrective action be taken, what will it be and which actor will take that action? You cannot try and promote the member indefinitely as then you increase the risk of running the entire etcd cluster with zero tolerance to yet another failure.

Reply:
> Is it possible that backup-restore container fails to promote a member 

`MemberPromote` call can be failed ....it is possible. 

> even after several retries?

I'm not sure even after several retries `MemberPromote` can be failed. 

> t what point a corrective action be taken, what will it be and which actor will take that action? You cannot try and promote the member indefinitely as then you increase the risk of running the entire etcd cluster with zero tolerance to yet another failure.

Do you have anything on your mind .... what action can be taken or when ..... but I would suggest that’s also depends on error we are getting while promoting the member.

Reply:
> You cannot try and promote the member indefinitely as then you increase the risk of running the entire etcd cluster with zero tolerance to yet another failure.

Same is true for the scale-up case. Any suggestion?